task_2/main.cpp

```diff
@@ -0,0 +1,197 @@
+#include <iostream>
+#include <typeinfo>
+#include <exception>
+
+using std::cout;
+using std::endl;
+using std::string;
+
+struct ICloneable {
+	virtual ICloneable* clone() const = 0;
+	virtual ICloneable* move() = 0;
+	virtual ~ICloneable() { }
+};
+
+template <typename T>
+struct ValueHolder : ICloneable {
+	T data_;		
+	ValueHolder(const T & arg) : data_ {arg} {}
+	ValueHolder * clone() const { 
+		return new ValueHolder(this->data_); 
+	}
+	ValueHolder * move() { 
+		return this; 
+	}
+};
+
+class Any {
+	ICloneable * ptr;			// pointer ICloneable * of base class can contain any type of it template child:
+						// ValueHolder<bool>, ValueHolder<int> or ValueHolder<double>
+public:
+	Any() : ptr {0} {}
+
+	template <typename T>
+	Any(const T & arg) {
+		string type = typeid(T).name();
+		if (type != "i" || type != "d" || type != "b")
+			ptr = 0;
```

Comment:
Name definitions are compiler specific. Consider using type != typeid(int).name() and so on. This shall both improve clarity and make program more adaptive to implementation specific changes.

Reply:
agree with it.
improved